receiver/prometheusremotewritereceiver/receiver.go

```diff
@@ -0,0 +1,243 @@
+
+		metricName := finaName(ts.Labels)
+		pm := pmetric.NewMetric()
+		pm.SetName(metricName)
+		rec.logger.Debug("Metric name", zap.String("metric_name", pm.Name()))
+
+		match := reg.FindStringSubmatch(metricName)
+		metricsType := ""
+		if len(match) > 1 {
+			lastSuffixInMetricName := match[len(match)-1]
+			if IsValidSuffix(lastSuffixInMetricName) {
+				metricsType = lastSuffixInMetricName
+				if len(match) > 2 {
+					secondSuffixInMetricName := match[len(match)-2]
+					if IsValidUnit(secondSuffixInMetricName) {
+						pm.SetUnit(secondSuffixInMetricName)
+					}
+				}
+			} else if IsValidUnit(lastSuffixInMetricName) {
+				pm.SetUnit(lastSuffixInMetricName)
+			}
+		}
+		rec.logger.Debug("Metric unit", zap.String("metric name", pm.Name()), zap.String("metric_unit", pm.Unit()))
+
+		for _, s := range ts.Samples {
+			ppoint := pmetric.NewNumberDataPoint()
+			ppoint.SetDoubleValue(s.Value)
+			ppoint.SetTimestamp(pcommon.NewTimestampFromTime(time.Unix(0, s.Timestamp*int64(time.Millisecond))))
+			if ppoint.Timestamp().AsTime().Before(timeThreshold) {
+				rec.logger.Debug("Metric older than the threshold", zap.String("metric name", pm.Name()), zap.Time("metric_timestamp", ppoint.Timestamp().AsTime()))
+				continue
+			}
+			for _, l := range ts.Labels {
+				labelName := l.Name
+				if l.Name == "__name__" {
+					labelName = "key_name"
+				}
+				ppoint.Attributes().PutString(labelName, l.Value)
+			}
+			if IsValidCumulativeSuffix(metricsType) {
+				pm.SetEmptySum()
+				pm.Sum().SetIsMonotonic(true)
+				pm.Sum().SetAggregationTemporality(pmetric.MetricAggregationTemporalityCumulative)
+				ppoint.CopyTo(pm.Sum().DataPoints().AppendEmpty())
+			} else {
+				pm.SetEmptyGauge()
+				ppoint.CopyTo(pm.Gauge().DataPoints().AppendEmpty())
+			}
+			rec.logger.Debug("Metric sample",
+				zap.String("metric_name", pm.Name()),
+				zap.String("metric_unit", pm.Unit()),
+				zap.Float64("metric_value", ppoint.DoubleValue()),
+				zap.Time("metric_timestamp", ppoint.Timestamp().AsTime()),
+				zap.String("metric_labels", fmt.Sprintf("%#v", ppoint.Attributes())),
+			)
+		}
+		pilm := pmetric.NewScopeMetrics()
+		pm.CopyTo(pilm.Metrics().AppendEmpty())
+		pilm.CopyTo(prm.ScopeMetrics().AppendEmpty())
+		prm.CopyTo(pms.ResourceMetrics().AppendEmpty())
+	}
+
+	metricCount := pms.ResourceMetrics().Len()
+	dataPointCount := pms.DataPointCount()
+	if metricCount != 0 {
+		err = rec.nextConsumer.ConsumeMetrics(ctx, pms)
+	}
+	rec.obsrecv.EndMetricsOp(ctx, receiverFormat, dataPointCount, err)
+	w.WriteHeader(http.StatusAccepted)
+}
+
+// Shutdown - remote write
+func (rec *PrometheusRemoteWriteReceiver) Shutdown(context.Context) error {
+	var err = component.ErrNilNextConsumer
+	rec.stopOnce.Do(func() {
+		err = rec.server.Close()
+		rec.shutdownWG.Wait()
+	})
+	return err
+}
+
+func finaName(labels []prompb.Label) (ret string) {
+	for _, label := range labels {
+		if label.Name == "__name__" {
+			return label.Value
+		}
+	}
+	return "error"
```

Comment:
Return an instance of `error` and handle it properly by the invocation logic.

Reply:
done